horace_core/sqw/@PixelData/PixelData.m

```diff
@@ -0,0 +1,172 @@
+classdef PixelData < matlab.mixin.SetGet
+% PixelData Provides an interface for access to pixel data
+%
+%   This class provides getters and setters for each data column in an SQW
+%   pixel array. You can access the data using the attributes listed below or
+%   using Matlab's get(obj, 'attr') and set(obj, 'attr', value) methods.
+%
+%
+% Usage
+%   >> pix_data = PixelData(data)
+%   >> signal = pix_data.signal;
+%
+%  or equivalently:
+%
+%   >> pix_data = PixelData()
+%   >> pix_data.data = data;
+%   >> signal = get(pix_data, 'signal');
+%
+% Attributes:
+%   data           The raw pixel data
+%   coordinates    Get/set the coords in projection axes of the pixel data (4 x n array)
+%   irun           The run index the pixel originated from (1 x n array)
+%   idet           The detector group number in the detector listing for the pixels (1 x n array)
+%   ienergy        The energy bin numbers (1 x n array)
+%   signals        The signal array (1 x n array)
+%   errors         The errors on the signal array (variance i.e. error bar squared) (1 x n array)
+%   num_pixels     The number of pixels in the data block
+
+properties (Access=private)
+    PIXEL_BLOCK_COLS_ = 9;
+    data_ = zeros(9, 0);
+end
+properties (Dependent)
+    % Returns the full raw pixel data block (9 x n array)
+    data;
+
+    % Returns the coordinates of the pixels in the projection axes, i.e.: u1,
+    % u2, u3 and dE (4 x n array)
+    coordinates;
+
+    % The run index the pixel originated from (1 x n array)
+    irun;
+
+    % The detector group number in the detector listing for the pixels (1 x n array)
+    idet;
+
+    % The energy bin numbers (1 x n array)
+    ienergy;
+
+    % The signal array (1 x n array)
+    signals;
+
+    % The errors on the signal array (variance i.e. error bar squared) (1 x n array)
+    errors;
+
+    % The number of pixels in the data block
+    num_pixels;
+end
+
+methods
+
+    function obj = PixelData(data)
+        % Construct a PixelData object from the given data. Default
+        % construction initialises the underlying data as an empty (9 x 0)
+        % array.
+        %
+        % Arguments:
+        %   data    A 9 x n matrix, where each row corresponds to a pixel and
+        %           the columns correspond to the following:
+        %             col 1: u1
+        %             col 2: u2
+        %             col 3: u3
+        %             col 4: u4
+        %             col 5: irun
+        %             col 6: idet
+        %             col 7: ienergy
+        %             col 8: signals
+        %             col 9: errors
+        %
+        if nargin == 1
```

Comment:
what if `nargin>1` ?

Reply:
You get a `Too many input arguments` error. We have the if statement since we still want to allow for zero input arguments (an empty `PixelData` block) - Mathworks also recommend creating a constructor that allows zero input args https://uk.mathworks.com/help/matlab/matlab_oop/class-constructor-methods.html.